certification/internal/shell/has_minimal_vulns.go

```diff
@@ -1,16 +1,96 @@
 package shell
 
 import (
+	"bytes"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+
 	"github.com/redhat-openshift-ecosystem/openshift-preflight/certification"
-	"github.com/redhat-openshift-ecosystem/openshift-preflight/certification/errors"
+	"github.com/redhat-openshift-ecosystem/openshift-preflight/certification/internal/utils/file"
 	"github.com/sirupsen/logrus"
 )
 
+const (
+	ovalFilename = "rhel-8.oval.xml.bz2"
+	ovalUrl      = "https://www.redhat.com/security/data/oval/v2/RHEL8/"
+	reportFile   = "vuln.html"
+)
+
 type HasMinimalVulnerabilitiesCheck struct{}
 
 func (p *HasMinimalVulnerabilitiesCheck) Validate(image string, logger *logrus.Logger) (bool, error) {
-	return false, errors.ErrFeatureNotImplemented
+
+	ovalFileUrl := fmt.Sprintf("%s%s", ovalUrl, ovalFilename)
+
+	tempdir, err := os.MkdirTemp(os.TempDir(), "oval-*")
+	if err != nil {
+		logger.Error("Unable to create temp dir", err)
+		return false, err
+	}
+	logger.Debugf("Oval file dir: %s", tempdir)
+	defer os.RemoveAll(filepath.Dir(tempdir))
+
+	// TODO claen up the names
+	ovalFilePath := filepath.Join(tempdir, ovalFilename)
+	logger.Debugf("Oval file path: %s", ovalFilePath)
+
+	err = file.DownloadFile(ovalFilePath, ovalFileUrl)
+	if err != nil {
+		logger.Error("Unable to download Oval file", err)
+		return false, err
+	}
+	// get the file name
+	r := regexp.MustCompile(`(?P<filename>.*).bz2`)
+	ovalFilePathDecompressed := filepath.Join(tempdir, r.FindStringSubmatch(ovalFilename)[1])
+
+	err = file.Unzip(ovalFilePath, ovalFilePathDecompressed)
+	if err != nil {
+		logger.Error("Unable to unzip Oval file: ", err)
+		return false, err
+	}
+
+	numOfVulns, err := numberOfVulnerabilities(image, ovalFilePathDecompressed, logger)
+	if err != nil {
+		return false, err
+	}
+
+	logger.Debugf("The number of found vulnerabilities: %d", numOfVulns)
+	if numOfVulns > 0 {
+		return false, nil
+	}
+
+	return true, nil
 }
+
+func numberOfVulnerabilities(image string, ovalFilePathDecompressed string, logger *logrus.Logger) (int, error) {
+
+	cmd := exec.Command("oscap-podman", image, "oval", "eval", "--report", reportFile, ovalFilePathDecompressed)
+	var out bytes.Buffer
+	cmd.Stdout = &out
+	var stderr bytes.Buffer
+	cmd.Stderr = &stderr
+	err := cmd.Run()
+
+	if err != nil {
+		logger.Error("unable to execute oscap-podman on the image: ", cmd.Stderr, reportFile)
+		return 0, err
+	}
+
+	path, err := os.Getwd()
+	if err != nil {
+		logger.Error("unable to get the current directory: ", err)
+	}
+
+	logger.Debugf("The path to vulnerability report: %s/%s ", path, reportFile)
+
+	r := regexp.MustCompile("Definition oval:com.redhat.*: true")
+	matches := r.FindAllStringIndex(string(out.String()), -1)
+	return len(matches), nil
+}
+
```

Comment:
Should this be a struct method on `HasMinimalVulnerabilitiesCheck`?

Reply:
hmm... it could be.. I see it more of a helper function, specific to this check.That's why I contained it to this type. Do you see it as a generic function?

Reply:
I'm currently parsing this as a generic helper function _because_ it's not bound to the struct. That is - I would expect any other check to be able to call this function the way it's written currently, so long as the caller is in the same package. 

I would bind it as an internal struct method (i.e. unexported, but still a method on `HasMinimalVulnerabilitiesCheck`) in a case where other checks won't have to use it (e.g. it's specific to this check).

Was your goal to write it so that only this check can use this function? Write now, I'm reading this as anything in the `shell` package can use this function. Could be wrong!

Reply:
Understood. Yes in this case, this must be bound to this struct. Thanks again :)